### PR TITLE
[ENH] Add `fit_kwargs` to Prophet

### LINKED ISSUE #2344: [ENH] Pass `pmdarima` and `prophet` models' `fit_params` to constructor
Native (not `sktime`) `pmdarima` models `ARIMA` and `AutoARIMA`, and `Prophet` (in`prophet`, not `sktime`) have some parameters which can be passed via `fit_params` (natively in `pmdarima` or `prophet`).
Users should be able to pass those parameters through the constructor of the `sktime` interface classes.
We should:

* check whether we are missing any parameters in the constructors that can be passed as `fit_params`. (if none are missing, we do not need to do the second step for that class) 
    - [ ] `Prophet`
    - [ ] `ARIMA`
    - [ ] `AutoARIMA` 
* if yes, add them to the constructor and ensure they are passed on to the interfaced model. The recipe is as below.
    - [ ] `Prophet`
    - [ ] `ARIMA`
    - [ ] `AutoARIMA` 
 
Good first issue since recipe is available.

---
Recipe for interfacing models with `fit_params` or any other `fit` args, in sktime:

* pass the parameter(s) to the constructor (the `__init__` method), using explicit parameter names and default, not `**kwargs`
* write it to `self` in the constructor
* in `_fit`, read it from `self` and pass it to the wrapped model's `fit` method
* ensure docstring is complete and defaults documented (optimally same as in wrapped estimator)

This allows to expose all parameter choices via the scikit-learn compatible interface (which relies on the constructor to expose all parameter choices).

For an example of this, you can look at `smoothing_level` in `ExponentialSmoothing`.

### PR DESCRIPTION
**Pull request by tpvasconcelos**
#### Reference Issues/PRs

Fixes #2344 

#### What does this implement/fix? Explain your changes.

Adds a dict `fit_kwargs` parameter to the Prophet forecaster constructor. This dictionary is then passed down to the `prophet.forecaster.Prophet.fit` method.

#### Does your contribution introduce a new dependency? If yes, which one?

No

#### Did you add any tests for the change?

Added a test that checks whether the correct kwargs are passed down to the `prophet.forecaster.Prophet.fit` method.

#### Any other comments?

No

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

File: sktime/forecasting/fbprophet.py [hunk 1]
```diff
@@ -162,6 +165,7 @@ def __init__(
         self.freq = freq
         self.add_seasonality = add_seasonality
         self.add_country_holidays = add_country_holidays
+        self.fit_kwargs = fit_kwargs or {}
```
Comment (fkiraly) [inline sktime/forecasting/fbprophet.py:168]: it is important that - in adherence with `sklearn` interface specifications, what is written to `self` is always identical with the `__init__` arg.

If you want a case distinction, you could use `self._fit_kwargs`.

**Comment by tpvasconcelos** [comment]
> * could you add a `get_test_params` parameter set that involves `fit_params`?

@fkiraly Prophet does not document which parameters are accepted here. These could also change depending on the selected Stan backend. In this case, I would say that there are no interesting cases to add to `get_test_params`. WDYT?

File: sktime/forecasting/fbprophet.py (hunk 1 above)
Comment (tpvasconcelos) [reply sktime/forecasting/fbprophet.py]: Good catch! I was not aware of the specification. Please review my changes

**Comment by fkiraly** [comment]
> I would say that there are no interesting cases to add to `get_test_params`. WDYT?

Well, that's odd - if we do not know a single example, are we sure it makes sense to add this as a parameter of the estimator?

**Comment by tpvasconcelos** [comment]
> Well, that's odd - if we do not know a single example, are we sure it makes sense to add this as a parameter of the estimator?

An example would be `prophet_forecaster.fit(df, seed=1234, algorithm="Newton")`. My point is that I don't think this is a **relevant** example, for the reasons I shared above and in my other comment. 

But maybe you can help me better understand what `get_test_params` is used for.

**Comment by fkiraly** [comment]
> But maybe you can help me better understand what `get_test_params` is used for.

this is used purely in testing, to generate a set of instances so that the estimator - or the interface code - receives good coverage, see last points in
https://www.sktime.net/en/latest/developer_guide/add_estimators.html#how-to-use-sktime-extension-templates

So, if we add the forward dict, we should add an arbitrary example where it is not `None`, so we test its forwarding.

**Comment by tpvasconcelos** [comment]
Got it! Then we can assume that the default stan backend will be used by prophet (cmdstanpy) and that these parameters will be passed down to [cmdstanpy.CmdStanModel.optimize](https://mc-stan.org/cmdstanpy/api.html#cmdstanpy.CmdStanModel.optimize) and [cmdstanpy.CmdStanModel.sample](https://mc-stan.org/cmdstanpy/api.html#cmdstanpy.CmdStanModel.sample)

File: sktime/forecasting/fbprophet.py [hunk 2]
```diff
@@ -233,5 +237,6 @@ def get_test_params(cls, parameter_set="default"):
             "daily_seasonality": False,
             "uncertainty_samples": 10,
             "verbose": False,
+            "fit_kwargs": {"seed": 12345},
```
Comment (tpvasconcelos) [inline sktime/forecasting/fbprophet.py:240]: @fkiraly like this?

Comment (fkiraly) [reply sktime/forecasting/fbprophet.py:240]: yes

File: sktime/forecasting/fbprophet.py [hunk 3]
```diff
@@ -41,6 +41,8 @@ class Prophet(_ProphetAdapter):
         Dict with args for Prophet.add_country_holidays().
         Dict can have the following keys/values:
             country_name: Name of the country, like 'UnitedStates' or 'US'
+    fit_kwargs: dict or None, default=None
+        Dict with args for Prophet.fit().
```
Comment (fkiraly) [inline sktime/forecasting/fbprophet.py:45]: perhaps explain better that these do not include data arguments.
We should also add what we know about valid values.

File: sktime/forecasting/fbprophet.py [hunk 4]
```diff
@@ -138,6 +140,7 @@ def __init__(
         freq=None,
         add_seasonality=None,
         add_country_holidays=None,
+        fit_kwargs=None,
```
Comment (fkiraly) [inline sktime/forecasting/fbprophet.py:143]: We should add it at the end, because here it is a breaking change that would need deprecation.
Why: suppose a user uses args, not kwargs, in the constructor.

File: sktime/forecasting/base/adapters/_fbprophet.py [hunk 5]
```diff
@@ -116,11 +116,12 @@ def _fit(self, y, X, fh):
             df["cap"] = self.growth_cap
             df["floor"] = self.growth_floor
 
+        fit_kwargs = self.fit_kwargs or {}
```
Comment (yarnabrina) [inline sktime/forecasting/base/adapters/_fbprophet.py:119]: Is there any place where it's ensured that all subclasses will have this attribute? If not, this is going to fail with AttributeError. I'll suggest to use hasattr or getattr to be safe.

Comment (fkiraly) [reply sktime/forecasting/base/adapters/_fbprophet.py:119]: yes, agreed - it also interacts with resolving the clash with the `sklearn` interface (write to `self` and do not change), so both should be addressed

Comment (fkiraly) [reply sktime/forecasting/base/adapters/_fbprophet.py:119]: > @fkiraly could you elaborate on this one?

Not changing the init params written to self is a must, for `sklearn` interface compatibility.

What I was saying, @yarnabrina's request relates to the same param affected by my comment related to this requirement, `fit_params`, so whatever the change to address it, it should also address the `sklearn` compatibility issue.

Comment (tpvasconcelos) [reply sktime/forecasting/base/adapters/_fbprophet.py:119]: @yarnabrina Thanks for the input!

Please take a look at this draft PR on my fork: https://github.com/tpvasconcelos/sktime/pull/1

I think that this tackles all of the issues you mentioned. 

I had a look at your `_GeneralisedStatsForecastAdapter` and although I think this is a very good example where the adapter pattern makes sense (since you have at least 7 concrete forecaster implementations from this base), it does not apply to Prophet since prophet only has one implementation and the adapter class currently 100% depends on its concrete implementation. The pattern that you exemplified in `_GeneralisedStatsForecastAdapter` can also be seen in all other adaptors (pmdarima, pytorch, tbats, ...). Prophet is the ugly outlier here.

Comment (tpvasconcelos) [reply sktime/forecasting/base/adapters/_fbprophet.py:119]: @yarnabrina 

If you really don't like this approach and [really really want](https://www.youtube.com/watch?v=gJLIiF15wjQ) an adapter class for Prophet, then I would suggest doing the same thing that was done with the `_TbatsAdapter` and move all logic to the adapter class and only define `self._ModelClass` (via `_create_model_class()`) in the concrete implementation... But, for the reasons I mentioned above, I don't really see how this is a good pattern for Prophet and it would just add unnecessary complexity to an already quite large codebase 😬

Comment (yarnabrina) [reply sktime/forecasting/base/adapters/_fbprophet.py:119]: Take a look at #5592 please. That is a PR which is expecting the prophet adapter. Also, even if not in current sktime codebase (I'm not certain, I don't use it much myself) someone using sktime may have inherited it in their codebase and added custom functionalities. It's not a public class, and there's no promise of it henceforth, but I personally don't think removal of the adapter is necessary.

This is just my personal opinion as of now, and it'snot a very strong one. If you feel strongly that it's unnecessary, may be you can create a new design issue and ask for others opinions? You can ask on discord as well. Either way, removal will be beyond scope of this PR for sure.

**Comment by tpvasconcelos** [comment]
@fkiraly Could you take a look at the current changes and let me know if there's anything that you'd still like me to address? Thanks in advance!

**Review by fkiraly — APPROVED** [review]
Yes, I think my change requests are addressed.

File: sktime/forecasting/fbprophet.py [hunk 6]
```diff
@@ -111,6 +111,10 @@ class Prophet(_ProphetAdapter):
     stan_backend: str or None, default=None
         str as defined in StanBackendEnum. If None, will try to
         iterate over all available backends and find the working one.
+    fit_kwargs: dict or None, default=None
```
Comment (fkiraly) [inline sktime/forecasting/fbprophet.py:114]: It would really be better to give a reference or link or list of possible parameters, or this will not be useful except for users who know prophet under the hood. Not blocking though.

Comment (tpvasconcelos) [reply sktime/forecasting/fbprophet.py:114]: @fkiraly 
Unfortunately Prophet itself does not provide this list of possible parameters anywhere in their documentation or source code. It's up to you ofc but I don't think we should make this sktime's responsibility to document. Also, note again that the accepted parameters will depend on the Stan backend being used. And when taking the default `cmdstanpy` backend, there are more than 30 possible parameters... 😬 (see [here](https://mc-stan.org/cmdstanpy/api.html#cmdstanpy.CmdStanModel.optimize) and [here](https://mc-stan.org/cmdstanpy/api.html#cmdstanpy.CmdStanModel.sample))

**Comment by tpvasconcelos** [comment]
@fkiraly @yarnabrina thanks again for all the valuable input, attention to detail, and promptness in reviewing this! It's highly appreciated ❤️

File: sktime/forecasting/fbprophet.py (hunk 6 above)
Comment (fkiraly) [reply sktime/forecasting/fbprophet.py:114]: @tpvasconcelos, I would suggest you add a short version of this to the docstring then, precisely what you said - prophet does not document this, our best guess is the stan backend params, link and link. Slightly more docstring tone.

Will merge this for now for the release, we can look at the docstring separately.
